### ISSUE TITLE
`deleterows!` is relatively slow and does lots of allocations

### ISSUE THREAD

**Issue by dmitrijsc** [JuliaData/DataFrames.jl#1371]
Hello,

I've been working with a relatively large dataset and been trying to remove rows by index. I found `deleterows!` to be slow and making lots of allocation.

Started looking through the source code and also found it has been widely used throughout the package and decided optimize it at least a little bit.

Please find an example of using `setdiff` and selection by index instead of `deleterows!`.

```julia
using DataFrames
using BenchmarkTools
using StatsBase

function test_deleterows()

    number_of_rows = 1000000
    rows_to_delete = 10000
    invalid_rows = unique(sample(1:number_of_rows, rows_to_delete))
    data = DataFrame(A = rand(number_of_rows), B = rand(number_of_rows), C = rand(number_of_rows))
    deleterows!(data, invalid_rows)
    return true
end

function test_setdiff()

    number_of_rows = 1000000
    rows_to_delete = 10000
    invalid_rows = unique(sample(1:number_of_rows, rows_to_delete))
    data = DataFrame(A = rand(number_of_rows), B = rand(number_of_rows), C = rand(number_of_rows))
    data = data[setdiff(1:nrow(data), invalid_rows), :];
    return true
end


@btime test_deleterows()
@btime test_setdiff()
```

```julia
Main> @btime test_deleterows()
  370.658 ms (9813743 allocations: 203.44 MiB)
true

Main> @btime test_setdiff()
  206.651 ms (203 allocations: 89.72 MiB)
true
```

**Comment by nalimilan** [comment]
Thanks for the reproducible example. It turns out this it due to `sort` being type-unstable on Julia 0.6, which kills inference for most variables in `deleterows!`. Fortunately, this is fixed on 0.7:
```julia
julia> @btime test_deleterows()
  29.084 ms (12367 allocations: 53.89 MiB)
true

julia> @btime test_setdiff()
  223.806 ms (1710892 allocations: 107.68 MiB)
true
```

So we just have to wait for a few weeks. ;-)

**Comment by dmitrijsc** [comment]
Thanks! I will close the ticket.
